### ISSUE TITLE
Use `rustc` attrs in `dangling_pointers_from_temporaries` lint

### ISSUE THREAD

**Issue by Urgau** [rust-lang/rust#132281]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/132281)*

> > If you want to do it in this PR (otherwise in a follow-up), we could add an `rustc` attribute to those `as_ptr`/`as_mut_ptr` methods.
>
> You can follow https://github.com/rust-lang/rust/pull/132146/commits/afcb09b8c765bf2c5b1362b433725f57af51aaad and https://github.com/rust-lang/rust/pull/132146/commits/2a930d3818d1f4dcaf7a06ed7fc12c102268755d (from a PR of mine) with `check_applied_to_fn_or_method` instead, and something similar to this to check the attribute https://github.com/rust-lang/rust/blob/e454c45f1397e382cfb59f0b8970445c8efa875f/compiler/rustc_lint/src/ptr_nulls.rs#L49-L50

_Originally posted by @Urgau in https://github.com/rust-lang/rust/pull/128985#discussion_r1818068264_

**Comment by Urgau** [comment]
## Steps

 1. Create the attribute (symbol, register, and validation), let's say `rustc_as_ptr_function` (better name to be found)
 2. Apply the newly created attribute to the relevant `as_ptr`/`as_mut_ptr` methods to the types listed 
    https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L202-L219
 3. Use the newly created attribute instead of `is_interesting` fn, with something similar to this
    https://github.com/rust-lang/rust/blob/e454c45f1397e382cfb59f0b8970445c8efa875f/compiler/rustc_lint/src/ptr_nulls.rs#L49-L50

More context in https://github.com/rust-lang/rust/pull/128985#discussion_r1818068264

**Comment by gavincrawford** [comment]
@GrigorenkoPV sorry, didn't realize this your follow up issue. Apologies.
@rustbot release-assignment

**Comment by GrigorenkoPV** [comment]
> [@GrigorenkoPV](https://github.com/GrigorenkoPV) sorry, didn't realize this your follow up issue. Apologies. [@rustbot](https://github.com/rustbot) release-assignment

I, most likely, won't have a chance to work on it in the next few weeks, so you can claim it in the meantime if you want to.

**Comment by gavincrawford** [comment]
Alright, I'll give it a shot. @rustbot claim

**Comment by GrigorenkoPV** [comment]
> 3\. Use the newly created attribute instead of `is_interesting` fn

I think the actual place where the attribute checking will have to go is this:

https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L136

so that we trigger on methods not based on their name but rather based on them having this attribute.

The `is_interesting` function checks that the type that we have *owns* the allocation. i.e. it's a `Vec<T>`, but not a `&Vec<T>` or a `&[T]`.

Ideally this should probably be unified with `is_temporary_rvalue`, so that we can properly handle field projections and whatnot, but that sounds like a lot of pain and chances to get it wrong. Probably out of scope for now.

In the meantime I see 3 options for `is_interesting`:

1. Leave as is (probably rename to `owns_allocation` or something along those lines for clarity). But add `UnsafeCell` and other missing things whose pointer-producing methods aren't named `as_ptr`/`as_mut_ptr` but will now be marked with the attribute: https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L46-L48
2. Mark the interesting types with an attribute. Probably not the same attribute as the methods. Replace this part https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L213-L219
3. Try to deduce the type based on the method's signature. Should work for `String`, but will probably not work for `Vec<T>` or `Box<[T]>` as their `as_ptr` methods actually come from `[T]` through `Deref`.

---

In conclusion,
1. Create a new attribute (for methods)
2. Mark the already-checked methods with it, as well as the currently unchecked `[Sync]UnsafeCell::get()`
3. Replace https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L136 with a check for the attribute.
4. Rename `is_interesting` to something about owning the memory.
5. Optionally: create a new attribute for types, mark the types mentioned here https://github.com/rust-lang/rust/blob/2dece5bb62f234f5622a08289c5a3d1555cd7843/compiler/rustc_lint/src/dangling.rs#L213-L219 and replace this snippet with a check for the attribute.

If you have any questions, feel free to reach out to me or Urgau (I guess they will not mind either, considering they've marked this issue with https://github.com/rust-lang/rust/labels/E-mentor)

**Comment by gavincrawford** [comment]
Thanks for all the helpful resources! I've created a new trait to track functions that implement `as_ptr` or `as_ptr_mut`. I'm having an issue where, when I attempt to apply this newly created attribute to functions, in this case, `Vec`, the compiler can't find the symbol.
![Image](https://github.com/user-attachments/assets/6ead03d4-05dd-4b2f-b80b-95a211a2490f)
I don't see where all the other macros and getting imported from, and my LSP can't figure it out either, showing no definitions found. Any help would be appreciated.

I'll list the commit from my personal fork [here](https://github.com/gavincrawford/rust/commit/4be0525ab502bc0b42c2382e717e72eb26b574df), which I did mostly with inspiration from the two commits posted for reference by Urgau. Thanks again for the help!

**Comment by GrigorenkoPV** [comment]
> ![Image](https://github.com/user-attachments/assets/6ead03d4-05dd-4b2f-b80b-95a211a2490f) 

I think I know what the problem might be.

Currently a new version of rustc is built as follows:
1. An old preexisting version of the compiler (called bootstrap) builds std.
2. The same bootstrap compiler builds the new version of the compiler using the std from the previous step.
3. This new version of the compiler now builds std too.
4. Some other steps (not important for now).

The problem is that the preexisting bootstrap has no idea about this new attribute. Only the new compiler will. Solution? `#[cfg(bootstrap)]`. (I assume you're familiar with [the `cfg` mechanism](https://doc.rust-lang.org/rust-by-example/attribute/cfg.html)).

Effectively `bootstrap` cfg will be set during the step 1, but not during the step 3.
So to hide this new attribute during step 1, but not during step 3, you should use
```rust
#[cfg_attr(not(bootstrap), rustc_as_ptr)]
```

As for the IDE problems, https://rustc-dev-guide.rust-lang.org/building/suggested.html?highlight=vscode#suggested-workflows might help.
TLDR: `./x setup` can generate the configurations necessary for your IDE to launch the LSP with appropriate settings. It can also generate `config.toml` for you.

Damn, Urgau was faster.

**Comment by gavincrawford** [comment]
I believe I've finished up with the primary objectives, which are:
1. Create a new attribute to track functions that return pointers, instead of matching function names
2. Add this new attribute to `Vec`, `String`, `CString`, `MaybeUninit`, `Cell`, `Box`, and finally, the `UnsafeCell::get()` function
3. Rename `is_interesting` to something more concise
4. Change the implementation for the dangling pointer lint to match for the new attribute

I've decided to postpone the fifth step mentioned in [this comment[(https://github.com/rust-lang/rust/issues/132281#issuecomment-2452466742) further up in the thread, as this is my first contribution here and I'd like to make sure I'm not making any mistakes before bundling that many edits into a single PR.

If everything sounds good, I'll go ahead and make a PR for review. My branch is linked [here](https://github.com/gavincrawford/rust/tree/as_ptr_attribute) if you'd like to take a look for any reason. Again, thank you @Urgau and @GrigorenkoPV for being so wonderfully helpful!

**Comment by Urgau** [comment]
The changes in your branch looks good. Feel free to create a PR (and assign it to me).

As for `in_interesting`/`is_own_allocation` I think we could simply replace it with something like this:
```rust
!matches!(ty, hir::TyKind::Ref(_, _))
```
which checks if a type isn't a reference.

**Comment by GrigorenkoPV** [comment]
> As for `in_interesting`/`is_own_allocation` I think we could/should simply replace it with something like this:
> 
> !matches!(ty.kind, hir::TyKind::Ref(_, _))
> which checks if a type isn't a reference.

I think then the lint will fire on things like `Box<&String>`, which is not what we want.

**Comment by Urgau** [comment]
Ah, right, forgot about that. I think that should handled with `expr_ty_adjusted` instead of `expr_ty` which handles auto-ref/auto-deref, but this becomes more trick so let's do that as a follow-up.

**Comment by GrigorenkoPV** [comment]
> Firstly, the `array` primitive type does not implement any `as_ptr` functions, as far as I know, so I didn't tag anything with the attribute. Am I missing anything?

I think it comes from slices and arrays just deref to those.

> Second, `MaybeUninit`'s `slice_as_ptr` method: I did not tag this function, as it internally references a regular `as_ptr` function, which has the attribute.

I don't think the attribute has any transitive effect. And the lint itself will only be matched against the code in the linted crate, so it won't peek into standard library's source code.

> Should the `slice_as_ptr` method also get the attribute, given that it _does_ return a raw pointer, despite the implementation?

Well, this function/method should probably be accounted for as well, but it is not a method on temporary (i.e. it is not `temp.slice_as_ptr()`, but rather rather `MaybeUninit::slice_as_ptr(temp)`), and thus requires some different handling, so I think this is out of scope for now.

**Comment by jjaemun** [comment]
Hi there, I would like to work on this issue as my first rustc contribution. Is it still available?

**Comment by gavincrawford** [comment]
PR #132732 is already merged, but I think there's definitely still some areas that were not covered in that PR. I'm not sure what is still relevant here, as it's been nearly two years since I wrote this code, but see the discussion above for certain things that we chose to put off for a while-- such as `::slice_as_ptr()`.

Personally, I think this issue might be stale, not necessarily because there's no more work to do, but because most of the specific improvements mentioned by this issue were fixed.

**Comment by jjaemun** [comment]
Hello Gavin. Thank you for your reply. Your point makes sense. Since I am quite new to this, I would like to ask for some advice -- should I try to build around the leftovers or try to find something else to do? I am happy with anything and eager to get started so I am all ears.

**Comment by gavincrawford** [comment]
Honestly, I might choose another issue that has an active mentor & clear objectives, but you definitely could try the remainder of this one. Nothing on this repo is particularly easy as a first PR. I had to search for weeks, and submit several rejected PRs, before I made this one work, and that was mostly because it was almost entire laid out in the original issue.

This stuff is hard, but there are always people willing to help! Whatever you choose to work on, we all appreciate your help, so never be afraid to ask anyone questions. Good luck!

**Comment by jjaemun** [comment]
Will do! Thanks a lot, really. I'll look for another good first issue, and hopefully we'll run into each other again in another discussion sometime!
